### PR TITLE
Fix IOS-1189:  image background color #fff vs #ffffff vs rgb(255,255,…

### PR DESCRIPTION
**Pull request by ChristineXYS**
…255,1)

Support color format rgb(255,255,255,1) by parsing red, green, blue and alpha.

### CODE REVIEW

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 1]
```diff
@@ -869,38 +869,69 @@ + (UIColor *)colorFromHexString:(NSString *)hexString
...
+        || hexString.length == 7 //#RRGGBB
+        || hexString.length == 9) //#AARRGGBB
+    {
+        if (![hexString hasPrefix:@"#"])
+        {
+            return nil;
+        }
+        if (hexString.length == 4)
+        {
+            NSString *red = [hexString substringWithRange:NSMakeRange(1, 1)];
```
Comment (asyncmind0) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: this should be extracted into a utility function ?

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 2]
```diff
@@ -869,38 +869,69 @@ + (UIColor *)colorFromHexString:(NSString *)hexString
...
-    green = ((rgbValue & 0xFF00) >> 8)/255.0;
-    blue = (rgbValue & 0xFF)/255.0;
-    if (hexString.length == 7)
-    {
-        alpha = 1.0;
+    CGFloat red = -1;
+    CGFloat green = -1;
+    CGFloat blue = -1;
+    CGFloat alpha = -1;
+    if (hexString.length == 4 //#RGB
```
Comment (StreethawkNick) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: Do you mean ?
var RGB = 4;
var RRGGBB = 7;
var AARRGGBB = 9;
try no magic numbers

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 3]
```diff
@@ -869,38 +869,69 @@ + (UIColor *)colorFromHexString:(NSString *)hexString
...
     {
-        alpha = ((rgbValue & 0xFF000000) >> 24)/255.0;
+        //rgb(255,255,255,1)
+        if ([hexString.lowercaseString hasPrefix:@"rgb("]
+            && [hexString.lowercaseString hasSuffix:@")"])
+        {
+            hexString = [hexString stringByReplacingOccurrencesOfString:@"rgb(" withString:@""];
+            hexString = [hexString stringByReplacingOccurrencesOfString:@")" withString:@""];
+            NSArray *arrayComponents = [hexString componentsSeparatedByString:@","];
+            if (arrayComponents.count >= 3)
```
Comment (StreethawkNick) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: magic number count >="3 and count == 4" what do they represent?

Comment (ChristineXYS) [reply StreetHawk/Classes/Core/Publish/SHUtils.m]: >= 3 means have R, G, B. 
>= 4 means have R, G, B, A.

File: StreetHawk/Classes/Core/Publish/SHUtils.m (hunk 2 above)
Comment (ChristineXYS) [reply StreetHawk/Classes/Core/Publish/SHUtils.m]: Yes. How to get rid of these number in logic?

**Review by djinoz — APPROVED** [review]

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 4]
```diff
@@ -863,44 +863,88 @@ - (NSString *)platformString
...
+}
+
++ (BOOL)isRGBA:(NSArray *)arrayComponents
+{
+    return arrayComponents.count == 4;
+}
+
 + (UIColor *)colorFromHexString:(NSString *)hexString
 {
+    const NSInteger RGB_COLOUR_CODE_LEN = 4;
```
Comment (asyncmind0) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: Make this global

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 5]
```diff
@@ -863,44 +863,88 @@ - (NSString *)platformString
 
 @implementation UIColor (SHExt)
 
++ (BOOL)isRGB:(NSArray *)arrayComponents
+{
+    return arrayComponents.count >= 3;
+}
```
Comment (asyncmind0) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: Use global constant instead of magic number

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 6]
```diff
@@ -863,44 +863,85 @@ - (NSString *)platformString
...
-    unsigned rgbValue = 0;
-    [scanner scanHexInt:&rgbValue];
-    CGFloat red, green, blue, alpha;
-    red = ((rgbValue & 0xFF0000) >> 16)/255.0;
-    green = ((rgbValue & 0xFF00) >> 8)/255.0;
-    blue = (rgbValue & 0xFF)/255.0;
-    if (hexString.length == 7)
-    {
-        alpha = 1.0;
+        if (![hexString hasPrefix:@"#"])
```
Comment (asyncmind0) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: extract function , Its not evident waht its doing here, so please extract into function with name that decribes what it does

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 7]
```diff
@@ -863,44 +863,85 @@ - (NSString *)platformString
...
+        else
+        {
+            alpha = ((rgbValue & 0xFF000000) >> 24)/255.0;
+        }
     }
     else
     {
-        alpha = ((rgbValue & 0xFF000000) >> 24)/255.0;
+        //rgb(255,255,255,1)
+        if ([hexString.lowercaseString hasPrefix:@"rgb("]
```
Comment (asyncmind0) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: extract function

**Review by asyncmind0 — CHANGES_REQUESTED** [review]

File: StreetHawk/Classes/Core/Publish/SHUtils.m [hunk 8]
```diff
@@ -863,44 +863,85 @@ - (NSString *)platformString
...
-        && hexString.length != 9) //AARRGGBB
-    {
-        return nil;
-    }
-    if (![hexString hasPrefix:@"#"])
+    CGFloat red = -1;
+    CGFloat green = -1;
+    CGFloat blue = -1;
+    CGFloat alpha = -1;
+    if (hexString.length == RGB_COLOUR_CODE_LEN + 1 //#RGB
```
Comment (asyncmind0) [inline StreetHawk/Classes/Core/Publish/SHUtils.m]: this checking should be isRGBString, isRGBAString isARGBAString

**Review by asyncmind0 — CHANGES_REQUESTED** [review]

**Review by asyncmind0 — APPROVED** [review]

**Pull request merged**